### ISSUE TITLE
Invalid attach limit value -1 cannot be added to CSINode object for "ebs.csi.aws.com"

### ISSUE THREAD

**Issue by jmhwang7** [kubernetes-sigs/aws-ebs-csi-driver#1361]
/kind bug

**What happened?**
We upgraded the ebs-csi-driver from version 1.4.0 to 1.10.0. We have an existing alert that tracks when more than 5 pods are stuck in ContainerCreating. This alert fired and upon running `kubectl describe` on the affected pods, we found the following events:
```
Events:
  Type     Reason              Age                 From                     Message
  ----     ------              ----                ----                     -------
  Normal   Scheduled           19m                 default-scheduler        Successfully assigned namespace_foo/pod_foo to ip-200-200-200-200.ap-southeast-1.compute.internal
  Warning  FailedMount         17m                 kubelet                  Unable to attach or mount volumes: unmounted volumes=[pgdata], unattached volumes=[user-volume root-volume pgconf-volume recover-volume pgdata primary-volume backup pgwal-volume dshm backrestrepo-volume]: timed out waiting for the condition
  
.... More FailedMount events

  Warning  FailedAttachVolume  48s (x10 over 19m)  attachdetach-controller  AttachVolume.Attach failed for volume "pvc-19d6d86d-ef49-43ad-a55b-14bd95969f9a" : rpc error: code = Internal desc = Could not attach volume "vol-03d1ad62aa5d7da79" to node "i-0b0a9b426ebfcd260": attachment of disk "vol-03d1ad62aa5d7da79" failed, expected device to be attached but was attaching
```

When we looked into the node, we found that there were a total of 31 pods with 1 volume each scheduled on the node (m5a.xlarge). 5 of those pods, the most recently scheduled on the node, were pending.

When investigating we found the following symptoms:
1. We had 2 affected nodes - both nodes had pods scheduled where their volumes had issues attaching and mounting. Both nodes had `CSINode` objects that looked like the following, and is missing an `allocatable` field which informs the scheduler how many volumes the node can accept:
```
kubectl get csinode/ip-200-200-200-200.ap-southeast-1.compute.internal -o yaml
apiVersion: storage.k8s.io/v1
kind: CSINode
metadata:
  ...
spec:
  drivers:
  - name: ebs.csi.aws.com
    nodeID: i-0b0a9b426ebfcd260
    topologyKeys:
    - topology.ebs.csi.aws.com/zone
```
2. Upon looking at kubelet logs (kubelet calls the `NodeGetInfo` function on the csi plugin when the plugin registers itself [here](https://github.com/kubernetes/kubernetes/blob/64ed9145452d2d1d324d2437566f1ea1ce76f226/pkg/volume/csi/csi_plugin.go#L135)), we see this error: `nodeinfomanager.go:566] Invalid attach limit value -1 cannot be added to CSINode object for "ebs.csi.aws.com"` (logged [here](https://github.com/kubernetes/kubernetes/blob/64ed9145452d2d1d324d2437566f1ea1ce76f226/pkg/volume/csi/nodeinfomanager/nodeinfomanager.go#L561)). This indicates that the attach limit, which is returned by nodeService.getVolumesLimit() is -1, which isn't correct - getVolumesLimit should return the `max_volumes_per_node` ([spec](https://github.com/container-storage-interface/spec/blob/master/spec.md#nodegetinfo)) , which should never be negative.

The scheduler keeps track of the currently attached volumes on a node. When scheduling a pod with a volume that is not already attached to the node, it takes the advertised volume limit of the node (which is returned by NodeGetInfo) and subtracts the currently attached volumes. If the result is a positive number, the scheduler schedules the pod on the node. If there is no limit (because the number returned by `NodeGetInfo` is negative), the scheduler always returns true ([code](https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/framework/plugins/nodevolumelimits/csi.go#L112)).

We have tracked down the issue to this change: https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1175 . This change starts returning the # of volumes that can be additionally attached to the node (it queries the number of volumes already attached) instead of the maximum number of volumes the instance can handle.

**What you expected to happen?**
nodeService.volumeLimits should return the maximum number of volumes the instance can handle. Kubernetes scheduler should not schedule pods with volumes on nodes that are at their volume limit.

**How to reproduce it (as minimally and precisely as possible)?**
1. Install aws-ebs-csi-driver v1.4.0.
2. Create a node with a volume limit (ex. m5a.xlarge).
3. Schedule as many pods as required to fill the node to its volume limit (ex: 25 pods with volumes on a m5a.xlarge)
4. Upgrade the aws-ebs-csi-driver to 1.10.0 (this restarts every ebs-csi-driver pod running on each node and kubelet will call `NodeGetInfo`). 
5. Schedule more pods with volumes on the same node - they should get scheduled and the pods will be stuck in ContainerCreating

**Anything else we need to know?**:

**Environment**
- Kubernetes version (use `kubectl version`): 1.21.7 and 1.21.9
- Driver version: 1.10

**Comment by ConnorJC3** [comment]
I think the issue here is that driver-managed volumes are being "double counted". If the node already has driver-managed volumes attached (for example, because of a driver restart or upgrade), then they should not be counted against the maxium number of volumes (but with the current code, they do).

Independently, there seems to be an off by 1 error with (at a minimum) `m5a.xlarge` instances, as even with the above described bug the math should arrive at 0 available volumes, not -1. We should fix this, but it also might be a good idea to clamp negative values to 0 just in case as well.

**Comment by ConnorJC3** [comment]
@jmhwang7 also, in the meantime, as a temporary fix you can manually supply the max number of volumes, with the helm parameter `node.volumeAttachLimit` (https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/charts/aws-ebs-csi-driver/values.yaml#L195), or the command line parameter `--volume-attach-limit` on the node.

**Comment by jmhwang7** [comment]
Thanks for the super fast response and the suggested temporary fix @ConnorJC3 ! We'll investigate setting that parameter.

**Comment by jmhwang7** [comment]
We deployed the temporary fix yesterday and so far so good! Every CSINode object has an allocatable count value: `kubectl get csinode -o custom-columns=NODE:.metadata.name,ALLOCATABLE:'.spec.drivers[0].allocatable.count' | grep none` is the command to run to find any problematic CSINode objects.

I found this [doc page](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/volume_limits.html) helpful for determining what number to set for the volume limit. All of our nodes are on the nitro system, have a single ENI, and have 1 ebs volume per node, which results in 26 volumes available for pods.

**Comment by ConnorJC3** [comment]
/remove-lifecycle stale
/add-lifecycle frozen

We need to keep this open until at the very least the "double counting" issue is fixed.

**Comment by ConnorJC3** [comment]
/reopen
/lifesycle frozen

Although we are now clamping the value to zero, we have not discovered the true root cause of this bug.

**Comment by jingxu97** [comment]
I see the fix PR is merged already. Is there any remaining issues for this? @ConnorJC3

**Comment by kon-angelo** [comment]
@ConnorJC3 we started experiencing this bug in our non-EKS cluster ([Gardener](https://github.com/gardener/gardener)) on AWS too. I would like to add our notes on this issue because, they weren't so apparent on my first read through of the issue.

> Although we are now clamping the value to zero, we have not discovered the true root cause of this bug.

While the effect that we are discussing here is correctly summarised as "double-counting", it doesn't seem to be exclusively a problem with the csi driver code. The counting depends on the instance metadata service which for `ebsN` volumes states ([ref](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-data-categories.html)):

> The virtual devices associated with any Amazon EBS volumes. Amazon EBS volumes are only available in metadata if they were present at launch time or when the instance was last started. The N indicates the index of the Amazon EBS volume (such as ebs1 or ebs2).

The reason why the current code works in the first place is because the metadata service kindly doesn't report *all* the attached volumes, but only those that were there when the instance launched. The CSI Driver pod itself can be updated and restarted "freely" as long as the metadata service replies with the correct response.

Of course, based on the docs description it is trivially easy to reproduce the issue in the driver by stopping and restarting an EC2 Instance while it was statefulsets. The result is that the `CSINode.Allocatable` field is consistently reduced  until it reaches 0. Also in the same experiment, you can verify that csi-driver restarts do not really have an effect on the outcome - it all depends on the metadata service.

In our case, the difference between a healthy node and and non-healthy one, was exactly that. The response of the metadata service started returning the `ebsN` volumes even though the **instance was never stopped**.  The launch time of the instance was much days before some of the last volume attachments, which indicated some issue with the AWS API. However, for the API we didn't find any apparent root cause and we plan to inquire this via support ticket.

Even though the problem seems to be on AWS API level, it is still an issue that the EBS-CSI driver can't discern between volumes that it manages and other volumes attached to the instance.

With that in mind:
- we can see that is possible for the attach limit of a node to be continuously degrade until it reaches 0
- the merged fix https://github.com/kubernetes-sigs/aws-ebs-csi-driver/pull/1542 while it does solve an issue, creates a different one. Now with an attach limit of 0, the nodes think they have unbounded `Allocatable`.  So:
   - the pods get scheduled into "full" nodes.
   - there is no indication like you would have in the "-1" case that the node has any issues. From OPs perspective this is a nightmare, because there is no error.
- the proposed solution to manually add the attach limit, is also a no-go for heterogenous clusters. 

tldr: 
- we suspect the root cause to be an AWS API issue
- howeveer the fact that the EBS-CSI-driver can't discern between EBS volumes that it manages is still a permanent issue.

**Comment by ConnorJC3** [comment]
/close

Closing this out as the primary issue has been solved. If you are running a CSI Driver and know in advance the number of slots that will be used by non-CSI volumes, we recommend passing this number to the EBS CSI driver via the `--reserved-volume-attachments` CLI flag (Helm parameter `node.reservedVolumeAttachments`), which will override the heuristic guess based on the number of attached volumes.
